quinn-proto/src/endpoint.rs

```diff
@@ -188,17 +193,46 @@ impl Endpoint {
         //
 
         let addresses = FourTuple { remote, local_ip };
-        if let Some(ch) = self.index.get(&addresses, &first_decode) {
-            return Some(DatagramEvent::ConnectionEvent(
-                ch,
-                ConnectionEvent(ConnectionEventInner::Datagram {
-                    now,
-                    remote: addresses.remote,
-                    ecn,
-                    first_decode,
-                    remaining,
-                }),
-            ));
+        if let Some(route_to) = self.index.get(&addresses, &first_decode) {
+            let event = DatagramConnectionEvent {
+                now,
+                remote: addresses.remote,
+                ecn,
+                first_decode,
+                remaining,
+            };
+            match route_to {
+                RouteDatagramTo::Incoming(incoming_idx) => {
+                    let incoming_buffer = &mut self.incoming_buffers[incoming_idx];
+                    let config = &self.server_config.as_ref().unwrap();
+
+                    if incoming_buffer
+                        .total_bytes
+                        .saturating_add(datagram_len as u64)
+                        < config.max_buffer_bytes_per_incoming
+                        && self
+                            .all_incoming_buffers_total_bytes
+                            .saturating_add(datagram_len as u64)
+                            < config.max_buffer_bytes_all_incoming
+                    {
+                        incoming_buffer.datagrams.push(event);
+                        incoming_buffer.total_bytes = incoming_buffer
+                            .total_bytes
+                            .saturating_add(datagram_len as u64);
+                        self.all_incoming_buffers_total_bytes += self
+                            .all_incoming_buffers_total_bytes
+                            .saturating_add(datagram_len as u64);
```

Comment:
This doubles the current value on each operation, irreversibly. Are we missing a unit test that would have caught that?

Reply:
:(

Reply:
I've reworked this logic to be based on `checked_add` now--I think it's a lot more intuitive and thus easier to read and less prone to these kids of programming mistakes and gives me more confidence that there's not some edge case.

A unit test that would've caught that would also have probably caught this bug in the first place when we were adding the Incoming API, given that their consequence is not connection failure outright but rather the 0-RTT feature appearing to engage successfully but not actually in practice achieving a request/response cycle in 1 round trip. I think that would be a good thing to do, but I'm not sure how easy it would be to do, and to what degree it would require essentially creating a basic network emulator. I might be really easy.

Reply:
Done.